### PR TITLE
GG-37926 Flaky test IgniteCachePutRetryAtomicSelfTest.testPutAll is fixed.

### PR DESCRIPTION
**Pull request by sergeyuttsel**
https://ggsystems.atlassian.net/browse/GG-37926

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateFuture.java [hunk 1]
```diff
@@ -486,6 +486,12 @@ private void sendDhtRequests(ClusterNode nearNode, GridCacheReturn ret, boolean
                 if (cntQryClsrs != null)
                     req.replyWithoutDelay(true);
 
+                if (updateRes.error() != null)
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateFuture.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateRequest.java [hunk 2]
```diff
@@ -502,54 +561,60 @@ final boolean isFlag(int mask) {
 
         switch (writer.state()) {
             case 4:
-                if (!writer.writeByte("flags", flags))
+                if (!writer.writeMessage("errs", errs))
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateRequest.java:549]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateFuture.java [hunk 3]
```diff
@@ -486,6 +486,12 @@ private void sendDhtRequests(ClusterNode nearNode, GridCacheReturn ret, boolean
...
+                    req.error(updateRes.error());
+
+                if (updateRes.failedKeys() != null)
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateFuture.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateFuture.java [hunk 4]
```diff
@@ -486,6 +486,12 @@ private void sendDhtRequests(ClusterNode nearNode, GridCacheReturn ret, boolean
...
+                    req.addFailedKeys(updateRes.failedKeys(), updateRes.error());
+
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateFuture.java:497]: Fixed

**Review by sk0x50 — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateRequest.java [hunk 5]
```diff
@@ -253,6 +260,43 @@ boolean hasResponse() {
...
+    @Override public IgniteCheckedException error() {
+        return errs != null ? errs.error() : null;
+    }
+
+    /**
+     * Sets update error.
+     *
+     * @param err Error.
+     */
+    public void error(IgniteCheckedException err) {
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicAbstractUpdateRequest.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicNearResponse.java [hunk 6]
```diff
@@ -125,6 +126,31 @@ public void addFailedKey(KeyCacheObject key, Throwable e) {
...
+
+        errs.addFailedKeys(keys, e);
+    }
+
+    /**
+     * Sets update error.
+     *
+     * @param err Error.
+     */
+    public void error(IgniteCheckedException err) {
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicNearResponse.java]: Fixed

**Pull request merged**